cmd/rpc.go

```diff
@@ -0,0 +1,287 @@
+	r.cmd.Flag("with-agent", "Match hosts with a certain agent").Short('A').StringsVar(&r.withAgent)
+	r.cmd.Flag("with-identity", "Match hosts with a certain configured identity").Short('I').StringsVar(&r.withIdentity)
+
+	r.cmd.Arg("agent", "Agent to call").Required().StringVar(&r.agent)
+	r.cmd.Arg("action", "Action to call").Required().StringVar(&r.action)
+	r.cmd.Arg("key=val", "Arguments to pass to agent").StringMapVar(&r.args)
+
+	return nil
+}
+
+func (r *rpcCommand) Configure() error {
+
+	err = commonConfigure()
+	if err != nil {
+		return err
+	}
+
+	if r.threaded {
+		cfg.Threaded = true
+	}
+	if r.target != "" {
+		cfg.MainCollective = r.target
+	}
+	if r.ttl != 0 {
+		cfg.TTL = r.ttl
+	}
+	if r.discoveryTimeout != 0 {
+		cfg.DiscoveryTimeout = r.discoveryTimeout
+	}
+	if r.discMethod != "" {
+		cfg.DefaultDiscoveryMethod = r.discMethod
+	}
+	if r.discOptions != nil {
+		cfg.DefaultDiscoveryOptions = r.discOptions
+	}
+	if r.publishTimeout != 0 {
+		cfg.PublishTimeout = r.publishTimeout
+	}
+	if r.connectionTimeout != 0 {
+		cfg.ConnectionTimeout = r.connectionTimeout
+	}
+
+	return nil
+}
+
+func (r *rpcCommand) rpcFilter() (*protocol.Filter, error) {
+	return client.NewFilter(
+		client.FactFilter(r.withFact...),
+		client.AgentFilter(r.withAgent...),
+		client.ClassFilter(r.withClass...),
+		client.IdentityFilter(r.withIdentity...),
+		client.CombinedFilter(r.sel...),
+		client.CompoundFilter(r.with...),
+	)
+}
+
+func (r *rpcCommand) Run(wg *sync.WaitGroup) (err error) {
+	defer wg.Done()
+
+	fw, err := choria.NewWithConfig(cfg)
+	if err != nil {
+		return err
+	}
+
+	// looks for json file
+	addl, err := ddl.Find(r.agent, cfg.LibDir)
+	if err != nil {
+		return err
+	}
+
+	rpc, err := rpcClient.New(fw, r.agent, rpcClient.DDL(addl))
+	if err != nil {
+		return err
+	}
+
+	var ctx context.Context
+	var cancel context.CancelFunc
+	if r.timeout != 0 {
+		ctx, cancel = context.WithTimeout(context.Background(), time.Duration(r.timeout)*time.Second)
+	} else {
+		ctx, cancel = context.WithCancel(context.Background())
+	}
+	defer cancel()
+
+	if !r.quiet {
+		fmt.Print("\nPerforming discovery...")
+	}
+	b := broadcast.New(fw)
+
+	filters, err := r.rpcFilter()
+	if err != nil {
+		return err
+	}
+	nodes, err := b.Discover(ctx, broadcast.Filter(filters))
+	if err != nil {
+		return err
+	}
+
+	if !r.quiet {
+		fmt.Printf("Number of nodes: %d\n\n", len(nodes))
+	}
+
+	if len(nodes) == 0 {
+		return fmt.Errorf("no nodes available to send requests to")
+	}
+
+	replies := make(chan jsonDisplay, len(nodes))
+
+	handler := rpcClient.ReplyHandler(func(reply protocol.Reply, rpcr *rpcClient.RPCReply) {
+		display := &jsonDisplay{}
+		err := json.Unmarshal([]byte(reply.Message()), display)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "failed to unmarshal incoming message from %s", reply.SenderID())
+			return
+		}
+		display.Agent = r.agent
+		display.Action = r.action
+		display.Sender = reply.SenderID()
+		replies <- *display
```

Comment:
this could cause trouble, if someone has 2 nodes with the same identity or if some new nodes appeared between discovery and request and a broadcast method is done you will get replied here that you do not expect and so overflow the channel leading to deadlock.

I'd set up a background routine to consume the replies channel and do what needs doing in parallel so it wont fill up.